src/plugins/data/public/search/session/session_service.ts

```diff
@@ -0,0 +1,243 @@
+    searchSessionRestorationInfoProvider: SearchSessionRestorationInfoProvider<ID> | undefined
+  ) {
+    this.searchSessionRestorationInfoProvider = searchSessionRestorationInfoProvider;
+  }
+
+  /**
+   * Used to track pending searches within current session
+   *
+   * @param searchDescriptor - uniq object that will be used to untrack the search
+   * @returns untrack function
+   */
+  public trackSearch(searchDescriptor: TrackSearchDescriptor): () => void {
+    if (!this.getSessionId())
+      throw new Error("Can't track search because there is no current session");
+    this.state.transitions.trackSearch(searchDescriptor);
+    return () => {
+      this.state.transitions.unTrackSearch(searchDescriptor);
+    };
+  }
+
+  public destroy() {
+    if (this.appChangeSubscription$) {
+      this.appChangeSubscription$.unsubscribe();
+    }
+    this.clear();
+    this.setSearchSessionRestorationInfoProvider(undefined);
+  }
+
+  /**
+   * Get current session id
+   */
+  public getSessionId() {
+    return this.state.get().sessionId;
+  }
+
+  /**
+   * Get observable for current session id
+   */
+  public getSession$() {
+    return this.state.state$.pipe(
+      startWith(this.state.get()),
+      map((s) => s.sessionId),
+      distinctUntilChanged()
+    );
+  }
+
+  /**
+   * Is current session already saved as SO (send to background)
+   */
+  public isStored() {
+    return this.state.get().isStored;
+  }
+
+  /**
+   * Is restoring the older saved searches
+   */
+  public isRestore() {
+    return this.state.get().isRestore;
+  }
+
+  /**
+   * Start a new search session
+   * @returns sessionId
+   */
+  public start() {
+    this.state.transitions.start();
+    return this.getSessionId()!;
+  }
+
+  /**
+   * Restore previously saved search session
+   * @param sessionId
+   */
+  public restore(sessionId: string) {
+    this.state.transitions.restore(sessionId);
+  }
+
+  /**
+   * Cleans up current state
+   */
+  public clear() {
+    this.state.transitions.clear();
+  }
+
+  private refresh$ = new Subject<void>();
+  /**
+   * Observable emits when search result refresh was requested
+   */
+  public onRefresh$ = this.refresh$.asObservable();
+
+  /**
+   * Request a search results refresh
+   */
+  public refresh() {
+    this.refresh$.next();
+  }
+
+  /**
+   * Request a cancellation of on-going search requests within current session
+   */
+  public async cancel(): Promise<void> {
+    const isStoredSession = this.state.get().isStored;
+    this.state.get().pendingSearches.forEach((s) => {
+      s.abort();
+    });
+    this.state.transitions.cancel();
+    if (isStoredSession) {
+      await this.sessionsClient.delete(this.state.get().sessionId!);
+    }
+  }
+
+  /**
+   * Save current session as SO to get back to results later
+   * (Send to background)
+   */
+  public async save(): Promise<void> {
+    const sessionId = this.getSessionId();
+    if (!sessionId) throw new Error('No current session');
```

Comment:
I don't think it is needed in this case, because this is more of `dev-facing` error for those how consumed our service in a wrong way. Users should never see those if service wire-up correctly.